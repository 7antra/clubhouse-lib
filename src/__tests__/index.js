/* @flow */

import Client from '../index';

class TestFactory {
  requests: Array<Request>;

  constructor(requests: Array<Request>) {
    this.requests = requests;
  }

  // eslint-disable-next-line class-methods-use-this
  makeRequest(url: string, method?: string, body?: Object): Request {
    return new Request(url, { method, body: JSON.stringify(body) });
  }

  executeRequest(request: Request): Promise<*> {
    this.requests.push(request);
    switch (request.method) {
      case 'GET':
      case 'HEAD':
      case 'DELETE':
        return Promise.resolve(
          new Response(JSON.stringify({}), {
            status: 200,
            headers: new Headers({ 'Content-Type': 'application/json' }),
          }),
        );
      default:
        return request.json().then(
          json =>
            new Response(JSON.stringify(json), {
              status: 200,
              headers: new Headers({ 'Content-Type': 'application/json' }),
            }),
        );
    }
  }
}

const makeClient = (factory: TestFactory) =>
  new Client(
    {
      baseURL: 'http://localhost:4001',
      version: 'beta',
    },
    factory,
  );

describe('#Client', () => {
  it('return a new instance with the correct defaults', () => {
    const client = Client.create('token value');
    expect(client.version).toBe('beta');
  });

  describe('.listProjects', () => {
    it('returns a list of projects with a clubhouse account', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.listProjects();

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.getProject', () => {
    it('requests a single project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.getProject(1234);

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.createProject', () => {
    it('creates a new project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.createProject({ name: 'test' });

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.updateProject', () => {
    it('updates a existing project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.updateProject(12, { name: 'test' });

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.deleteProject', () => {
    it('deletes a existing project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.deleteProject(1);

      expect(requests).toMatchSnapshot();
    });
  });
});
