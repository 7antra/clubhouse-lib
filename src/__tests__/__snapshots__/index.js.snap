// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Client .createProject creates a new project 1`] = `
Array [
  Body {
    "_abort": false,
    "_bytes": 15,
    "_raw": Array [
      Object {
        "data": Array [
          123,
          34,
          110,
          97,
          109,
          101,
          34,
          58,
          34,
          116,
          101,
          115,
          116,
          34,
          125,
        ],
        "type": "Buffer",
      },
    ],
    "agent": undefined,
    "auth": null,
    "body": "{\\"name\\":\\"test\\"}",
    "bodyUsed": true,
    "compress": true,
    "counter": 0,
    "follow": 20,
    "headers": Headers {
      "_headers": Object {},
    },
    "hostname": "localhost",
    "method": "POST",
    "path": "/api/beta/projects",
    "port": "4001",
    "protocol": "http:",
    "redirect": "follow",
    "size": 0,
    "timeout": 0,
    "url": "http://localhost:4001/api/beta/projects",
  },
]
`;

exports[`#Client .deleteProject deletes a existing project 1`] = `
Array [
  Body {
    "_abort": false,
    "_raw": Array [],
    "agent": undefined,
    "auth": null,
    "body": undefined,
    "bodyUsed": false,
    "compress": true,
    "counter": 0,
    "follow": 20,
    "headers": Headers {
      "_headers": Object {},
    },
    "hostname": "localhost",
    "method": "DELETE",
    "path": "/api/beta/projects/1",
    "port": "4001",
    "protocol": "http:",
    "redirect": "follow",
    "size": 0,
    "timeout": 0,
    "url": "http://localhost:4001/api/beta/projects/1",
  },
]
`;

exports[`#Client .getProject requests a single project 1`] = `
Array [
  Body {
    "_abort": false,
    "_raw": Array [],
    "agent": undefined,
    "auth": null,
    "body": undefined,
    "bodyUsed": false,
    "compress": true,
    "counter": 0,
    "follow": 20,
    "headers": Headers {
      "_headers": Object {},
    },
    "hostname": "localhost",
    "method": "GET",
    "path": "/api/beta/projects/1234",
    "port": "4001",
    "protocol": "http:",
    "redirect": "follow",
    "size": 0,
    "timeout": 0,
    "url": "http://localhost:4001/api/beta/projects/1234",
  },
]
`;

exports[`#Client .listProjects returns a list of projects with a clubhouse account 1`] = `
Array [
  Body {
    "_abort": false,
    "_raw": Array [],
    "agent": undefined,
    "auth": null,
    "body": undefined,
    "bodyUsed": false,
    "compress": true,
    "counter": 0,
    "follow": 20,
    "headers": Headers {
      "_headers": Object {},
    },
    "hostname": "localhost",
    "method": "GET",
    "path": "/api/beta/projects",
    "port": "4001",
    "protocol": "http:",
    "redirect": "follow",
    "size": 0,
    "timeout": 0,
    "url": "http://localhost:4001/api/beta/projects",
  },
]
`;

exports[`#Client .updateProject updates a existing project 1`] = `
Array [
  Body {
    "_abort": false,
    "_bytes": 15,
    "_raw": Array [
      Object {
        "data": Array [
          123,
          34,
          110,
          97,
          109,
          101,
          34,
          58,
          34,
          116,
          101,
          115,
          116,
          34,
          125,
        ],
        "type": "Buffer",
      },
    ],
    "agent": undefined,
    "auth": null,
    "body": "{\\"name\\":\\"test\\"}",
    "bodyUsed": true,
    "compress": true,
    "counter": 0,
    "follow": 20,
    "headers": Headers {
      "_headers": Object {},
    },
    "hostname": "localhost",
    "method": "PUT",
    "path": "/api/beta/projects/12",
    "port": "4001",
    "protocol": "http:",
    "redirect": "follow",
    "size": 0,
    "timeout": 0,
    "url": "http://localhost:4001/api/beta/projects/12",
  },
]
`;
